daemon/container_linux.go

```diff
@@ -737,27 +737,89 @@ func (container *Container) buildCreateEndpointOptions() ([]libnetwork.EndpointO
 	return createOptions, nil
 }
 
+func parseService(controller libnetwork.NetworkController, service string) (string, string, string) {
+	dn := controller.Config().Daemon.DefaultNetwork
+	dd := controller.Config().Daemon.DefaultDriver
+
+	snd := strings.Split(service, ".")
+	if len(snd) > 2 {
+		return strings.Join(snd[:len(snd)-2], "."), snd[len(snd)-2], snd[len(snd)-1]
+	}
+	if len(snd) > 1 {
+		return snd[0], snd[1], dd
+	}
+	return snd[0], dn, dd
+}
+
+func createNetwork(controller libnetwork.NetworkController, dnet string, driver string) (libnetwork.Network, error) {
+	createOptions := []libnetwork.NetworkOption{}
+	genericOption := options.Generic{}
+
+	// Bridge driver is special due to legacy reasons
+	if runconfig.NetworkMode(driver).IsBridge() {
+		genericOption[netlabel.GenericData] = map[string]interface{}{
+			"BridgeName":            dnet,
+			"AllowNonDefaultBridge": "true",
+		}
+		networkOption := libnetwork.NetworkOptionGeneric(genericOption)
+		createOptions = append(createOptions, networkOption)
+	}
+
+	return controller.NewNetwork(driver, dnet, createOptions...)
+}
+
 func (container *Container) AllocateNetwork() error {
 	mode := container.hostConfig.NetworkMode
+	controller := container.daemon.netController
 	if container.Config.NetworkDisabled || mode.IsContainer() {
 		return nil
 	}
 
-	var err error
+	var networkDriver string
+	service := container.Config.PublishService
+	networkName := mode.NetworkName()
+	canCreateNetwork := false
+	if mode.IsDefault() {
+		canCreateNetwork = true
+		if service != "" {
+			service, networkName, networkDriver = parseService(controller, service)
```

Comment:
I agree with @squaremo.  Services and networks should be distinct concepts.  Hiding the basic primitives of networking behind a higher level construct that some (but not all) people will want to use feels wrong to me.  Services as envisaged here are a good thing (and plenty of people will want to use them).  But there are people who will not, particularly those running at large scale.  So I do think there is a lot of value in including some equivalent of the originally discussed "docker run --net=foo ..." which can be used independently of services.

For example, what is the proposed approach for someone who for example wants to spin up 100 back-end containers for their application running at scale?  Do they need to give each one a unique service name?  And DNS gets filled up with an entry for every container?  

I¹m concerned that for people running at scale this is going to be very painful.  We have customers looking to deploy Calico on thousands of servers in their data center, who are looking seriously at containers for their workloads.  In a data center of that size you are talking 100,000 containers.  Giving each one a unique service name and an entry in DNS?  I¹m not so sure that's going to work well for them.

We also need to remember that there are some fairly major orchestration systems that orchestrate Docker today which have their own service discovery mechanisms.  We need to make sure the Docker API & UI is friendly to their needs too.